### DISCUSSION TITLE
non-decision time does not converge

### DISCUSSION THREAD

**Discussion by TabithaSteendam** [lnccbrown/HSSM#323] [Q&A]
Hi all,

I was trying to do a regression on a, v, and t but had convergence issues.

I then tried modeling a single participant in the simplest version and the t parameter does not converge. 
This is my code:
```
m = hssm.HSSM(
    data=data3,
    model="ddm",    
)

infer_m = m.sample(
    sampler="nuts_numpyro", chains=4, cores=1, draws=5000, tune=3000
)
```
This is what the data looks like (errors are -1):

![image](https://github.com/lnccbrown/HSSM/assets/70645340/2b1ee6d3-69a5-4f21-b4dc-143849321ee5)

And these are the traces:
![image](https://github.com/lnccbrown/HSSM/assets/70645340/c7d9b219-20ca-4476-bd8e-b1ba99cc0721)

I tried different participants and different datasets but it is always t that doesn’t converge. I tried setting a prior for t and also tried setting initval, but no success.

Does anyone have an idea why this is happening?

**Comment by AlexanderFengler** [comment]
There are three options that you can test without too much hassle. 

1. specify the prior on the non-decision time and set the `initval` parameter to something very small (e.g. `0.01`)
2. try to run a slice sampler instead --> the easiest way to do so is to simply set the `likelihood_kind` argument at initialization of the `HSSM` class to 'blackbox' which will automatically choose a slice sampler
3. try to run a slice sampler with the default likelihood --> this code `infer_m = m.sample(
    sampler=pm.Slice(model = m), chains=4, cores=1, draws=5000, tune=3000
)` should work, but let us know if it doesn't.

**Reply by TabithaSteendam** [reply]
I tried out the options but still run into problems. I wonder if they arise from package version problems 

- For the slice sampler` infer_m = m.sample( sampler=pm.Slice(model = m), chains=4, cores=1, draws=5000, tune=3000) ` I get the error:
Traceback (most recent call last):

  Cell In[5], line 1
    infer_m = m.sample( sampler=pm.Slice(model = m), chains=4, cores=1, draws=5000, tune=3000 )

  File ~\.conda\envs\HSSM\lib\site-packages\pymc\step_methods\compound.py:126 in __new__
    vars = model.value_vars

AttributeError: 'HSSM' object has no attribute 'value_vars'

- For the likelihood approx_differentiable I get the error:
Traceback (most recent call last):

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\link\vm.py:414 in __call__
    thunk()

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\graph\op.py:543 in rval
    r = p(n, [x[0] for x in i], o)

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\tensor\elemwise.py:796 in perform
    odat[...] = variable

ValueError: assignment destination is read-only


During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  Cell In[3], line 42
    infer_m = m.sample(

  File ~\.conda\envs\HSSM\lib\site-packages\hssm\hssm.py:343 in sample
    self._inference_obj = self.model.fit(inference_method=sampler, **kwargs)

  File ~\.conda\envs\HSSM\lib\site-packages\bambi\models.py:325 in fit
    return self.backend.run(

  File ~\.conda\envs\HSSM\lib\site-packages\bambi\backend\pymc.py:96 in run
    result = self._run_mcmc(

  File ~\.conda\envs\HSSM\lib\site-packages\bambi\backend\pymc.py:211 in _run_mcmc
    idata = pymc.sampling_jax.sample_numpyro_nuts(

  File ~\.conda\envs\HSSM\lib\site-packages\pymc\sampling\jax.py:623 in sample_numpyro_nuts
    init_params = _get_batched_jittered_initial_points(

  File ~\.conda\envs\HSSM\lib\site-packages\pymc\sampling\jax.py:247 in _get_batched_jittered_initial_points
    initial_points = _init_jitter(

  File ~\.conda\envs\HSSM\lib\site-packages\pymc\sampling\mcmc.py:1221 in _init_jitter
    model.check_start_vals(point)

  File ~\.conda\envs\HSSM\lib\site-packages\pymc\model.py:1697 in check_start_vals
    initial_eval = self.point_logps(point=elem)

  File ~\.conda\envs\HSSM\lib\site-packages\pymc\model.py:1732 in point_logps
    self.compile_fn(factor_logps_fn)(point),

  File ~\.conda\envs\HSSM\lib\site-packages\pymc\pytensorf.py:763 in __call__
    return self.f(**state)

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\compile\function\types.py:970 in __call__
    self.vm()

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\link\vm.py:418 in __call__
    raise_with_op(self.fgraph, node, thunk)

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\link\utils.py:535 in raise_with_op
    raise exc_value.with_traceback(exc_trace)

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\link\vm.py:414 in __call__
    thunk()

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\graph\op.py:543 in rval
    r = p(n, [x[0] for x in i], o)

  File ~\.conda\envs\HSSM\lib\site-packages\pytensor\tensor\elemwise.py:796 in perform
    odat[...] = variable

ValueError: assignment destination is read-only

I cannot change "odat" to writable and also cannot make a copy.

- The likelihood blackbox only progressed 0.05% in 24 hours on a small dataset of 4 participants. 


Is there a document that has pip freeze like overview of all the package versions that are required for a working HSSM so I can compare my versions?

**Comment by AlexanderFengler** [comment]
Some more context. 

The analytical likelihood we are using has some issues when the non-decision time is very close to the smallest rt in your data (which, looking at your rt column could be quite possible for a setting of `t` around `0.26` secionds). Gradients become unstable which translated to divergences in the context of NUTS.

This is actually something we learned and are trying to robustify, but for now you will sometimes run into that. The `approx_differentiable` likelihood in fact does not suffer from the same shortcoming, so you can also use that for `likelihood_kind`, but it is "approximate".

The solution I suggested to you goes for a sampler that instead doesn't make use of gradients, so the issue won't appear, but uses more "exact" likelihoods.

**Comment by TabithaSteendam** [comment]
Thanks for your explanation and advise! I'm going to try this
